### ISSUE TITLE
Generator.integers() always produces the same sample under specific circumstances

### ISSUE THREAD

**Issue by aleju** [numpy/numpy#14213]
I encountered a weird scenario, in which it is possible to induce a `numpy.random.Generator` to always produce the same integer value upon calling `integers()`, even though the internal state of `Generator.bit_generator.state` seemingly changes. This only happens if `integers()` is initially called a specific number of times, otherwise the sampled values will behave (pseudo-)randomly as expected. It also only affects the first sampled integer in case `size=(...)` is added to the call.

### Reproducing code example:

```python
import numpy as np
from copy import deepcopy
import sys, numpy; print(numpy.__version__, sys.version)

SEED = 1

BITGEN = np.random.SFC64
# BITGEN = np.random.PCG64
# BITGEN = np.random.Philox
N = 1
M = 100


def main():
    rng = np.random.Generator(BITGEN(SEED))

    # Issue only happens if N>0 *and* N is odd
    for _ in range(N):
        rng.integers(0, 2**31-1)

    ints, states = [], []

    for _ in range(4):
        # sampled value of "int_" differs if:
        # N==0, N is even, call is random() instead of integers()
        int_  = deepcopy(rng).integers(0, 2**31-1)
        state = deepcopy(rng.bit_generator.state)

        # Advance the RNG's internal state.
        # This *should* usually make sure that the next iteration's call
        # to integers() produces different samples.
        for _ in range(M):
            rng.uniform()

        ints.append(int_)
        states.append(state)

    for name, values in zip(["ints", "states"],
                            [ints, states]):
        print("")
        print(name)
        for value in values:
            print(value)


if __name__ == "__main__":
    main()
```

### Output

```bash
1.17.0 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21) 
[GCC 7.3.0]

ints
2138077778
2138077778
2138077778
2138077778

states
{'bit_generator': 'SFC64', 'state': {'state': array([10822012043237629067, 14489128587111033830,  3369327766797514304,
                         14], dtype=uint64)}, 'has_uint32': 1, 'uinteger': 4276155558}
{'bit_generator': 'SFC64', 'state': {'state': array([17974599381666177023, 10461407714979245550,  9510091932853551627,
                        114], dtype=uint64)}, 'has_uint32': 1, 'uinteger': 4276155558}
{'bit_generator': 'SFC64', 'state': {'state': array([15153555155425159765,  5979911724464602935,  7936913977336769722,
                        214], dtype=uint64)}, 'has_uint32': 1, 'uinteger': 4276155558}
{'bit_generator': 'SFC64', 'state': {'state': array([14366095984858144903, 18327215719122944750, 10053482158821344921,
                        314], dtype=uint64)}, 'has_uint32': 1, 'uinteger': 4276155558}
```

### What Affects the Issue

* Using a different bit generator for `BITGEN` has no effect.
* Changing `SEED` has no effect (different sampled number that keeps repeating).
* Changing `M` has no effect.
* `N` however is important. It has to be set to a value that is `>0` **and** odd. Otherwise the integers will differ between iterations of the loop.
* Calling `random()` instead of `integers()` makes the outputs differ between iterations.
* Adding `dtype="int32"` or `dtype="int64"` to the second call of `integers()` has no effect.
* Adding `size=(2,)` to the second call of `integers()` will produce for each iteration two values `(a, b)`, where `a` is the same for all iterations and b differs between all iterations. For `size=(3,)` producing tuples `(a, b, c)`, `a` stays the same for all iterations, while both `b` and `c` differ.

**Comment by rkern** [comment]
I believe everything is working as designed.

Note the `'has_uint32': 1, 'uinteger': 4276155558` parts of the state. All of these `BitGenerator`s natively generate `uint64`s. When we generate 32-bit integers (judged by the `low` and `high`; the `dtype='uint64'` doesn't matter if `high` is small enough; it only controls the type of the output, not the method by which the values are generated), we generate 2 at a time for efficiency. When you generate an odd number of 32-bit integers, you have one left over in the cache. You can do arbitrary amounts of random number generation that doesn't consume that cached value, then come back and generate another 32-bit integers: you will get the cached value first without advancing the `BitGenerator` state.

This is similar to how we worked with `RandomState.standard_normal()`. For the old `RandomState`, we used an algorithm that generated 2 normal variates at a time, so if you requested an odd number, you are left with a cached value left over.

**Comment by aleju** [comment]
I see, that sounds sensible. Thanks for the explanation.
When using `N=2` in the above script I ended up getting different sampled integers with each iteration, but the `has_uint32` and `uinteger` values still didn't change, so I ended up ignoring them and only compared the `state` dictionaries. Though looking at it again now I notice that `has_uint32` is `0` and not `1` when using `N=2`. Does that flag indicate whether the `uinteger` value is supposed to be used as a cached integer for the next call? So, setting `has_uint32` to `0` effectively removes all cache from the internal state?
